### PR TITLE
[SYCLomatic] Introduce an Attribute field in user define rule

### PR DESCRIPTION
**Pull request by yihwang-nv**
- Introduce an Attribute field in user define rule
- Introduce an Attribute kind `CalleeNameOnly` in user define rule

Signed-off-by: Wang, Yihan <yihan.wang@intel.com>

### CODE REVIEW

File: clang/test/dpct/replace_callee_name_only.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+- Rule: replace_callee_name_rule
+  Kind: API
+  Priority: Takeover
+  In: cub::DeviceScan::InclusiveSum
+  Out: cub::DeviceScan::InclusiveSum
+  Includes: []
+  Attribute:                                                                    
+    - Kind: CalleeNameOnly
```
Comment (AndyCHHuang) [inline clang/test/dpct/replace_callee_name_only.yaml]: Attribute:
    - CalleeNameOnly: yes

Comment (yihwang-nv) [reply clang/test/dpct/replace_callee_name_only.yaml]: done

File: clang/lib/DPCT/Rules.h [hunk 2]
```diff
@@ -166,6 +172,13 @@ struct llvm::yaml::MappingTraits<std::shared_ptr<MetaRuleObject::ClassMethod>> {
   }
 };
 
+template<>
+struct llvm::yaml::MappingTraits<MetaRuleObject::Attribute> {
+  static void mapping(llvm::yaml::IO &Io, MetaRuleObject::Attribute &Doc) {
+    Io.mapRequired("ReplaceCalleeNameOnly", Doc.ReplaceCalleeNameOnly);
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/Rules.h]: Optional

Comment (yihwang-nv) [reply clang/lib/DPCT/Rules.h]: done

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/test/dpct/replace_callee_name_only.yaml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+- Rule: rule0
+  Kind: API
+  Priority: Takeover
+  In: cub::DeviceReduce::Sum
+  Out: REPLACE_CALLEE_NAME_ONLY
```
Comment (zhimingwang36) [inline clang/test/dpct/replace_callee_name_only.yaml]: let's use a meaningful API here,  in case user want to build the lit case.

Comment (yihwang-nv) [reply clang/test/dpct/replace_callee_name_only.yaml]: done

File: clang/lib/DPCT/Rules.h [hunk 4]
```diff
@@ -71,6 +71,10 @@ class MetaRuleObject {
     std::string Out;
     ClassMethod() {}
   };
+  class Attribute {
```
Comment (ziranzha) [inline clang/lib/DPCT/Rules.h]: How about use "struct Attributes"

File: clang/lib/DPCT/CallExprRewriter.h [hunk 5]
```diff
@@ -1439,7 +1442,8 @@ class UserDefinedRewriterFactory : public CallExprRewriterFactoryBase {
 public:
   UserDefinedRewriterFactory(MetaRuleObject &R)
       : OutStr(R.Out), Includes(R.Includes),
-        HasExplicitTemplateArgs(R.HasExplicitTemplateArgs) {
+        HasExplicitTemplateArgs(R.HasExplicitTemplateArgs),
```
Comment (ziranzha) [inline clang/lib/DPCT/CallExprRewriter.h]: Consider move HasExplicitTemplateArgs into Attributes.

Comment (yihwang-nv) [reply clang/lib/DPCT/CallExprRewriter.h]: done

File: clang/lib/DPCT/Rules.h (hunk 4 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/Rules.h]: done

**Review by ziranzha — APPROVED** [review]

**Review by AndyCHHuang — APPROVED** [review]
LGTM

**Pull request merged**